textbook/chapter_1.qmd

```diff
@@ -0,0 +1,993 @@
+        max_tokens: 100,
+        top_p: 1,
+        frequency_penalty: 0,
+        presence_penalty: 0
+      })
+    });
+
+    if (!response.ok) {
+      throw new Error(`HTTP error! status: ${response.status}`);
+    }
+
+    const data = await response.json();
+    const assistantResponse = data.choices[0].message.content;
+
+    // Add assistant's response to conversation history
+    messages.push({ role: "assistant", content: assistantResponse });
+
+    return assistantResponse;
+  } catch (error) {
+    console.error("Error:", error);
+    throw error;
+  }
+}
+
+const container = html`<div id="chat-container">
+  <div class="highlight-line"></div>
+  <div id="chat-display"></div>
+  <textarea id="user-input" placeholder="Enter your message here"></textarea>
+  <button id="submit-button">Send Message</button>
+</div>`;
+
+const chatDisplay = container.querySelector('#chat-display');
+const userInput = container.querySelector('#user-input');
+const submitButton = container.querySelector('#submit-button');
+
+function createChatMessageHTML(content, isBoss) {
+  return html`
+    <div class="chat-message ${isBoss ? 'boss-message' : ''}">
+      <div class="chat-avatar" style="background-image: url('${isBoss ? '/img/environ_minister.jpg' : '/img/zephyr.jpg'}')"></div>
+      <div class="chat-message-content">
+        <div class="chat-sender-name">${isBoss ? 'Minister Reed' : 'Zephyr Eco-Luna'}</div>
+        <div>${content}</div>
+      </div>
+    </div>
+  `;
+}
+
+// Initial message
+chatDisplay.appendChild(createChatMessageHTML(first_message, true));
+
+submitButton.onclick = async () => {
+  const prompt = userInput.value;
+  if (prompt.trim() === '') return;
+
+  messageCount++;
+
+  // Display user's message
+  chatDisplay.appendChild(createChatMessageHTML(prompt, false));
+  userInput.value = '';
+  chatDisplay.scrollTop = chatDisplay.scrollHeight;
+
+  try {
+    const response = await sendOpenAIRequest(prompt);
+    chatDisplay.appendChild(createChatMessageHTML(response, true));
+  } catch (error) {
+    chatDisplay.appendChild(createChatMessageHTML("Error: " + error.message, true));
+  }
+
+  // Scroll to the bottom of the chat display
+  chatDisplay.scrollTop = chatDisplay.scrollHeight;
+};
+
+chatDisplay.scrollTop = chatDisplay.scrollHeight;
+
+return html`${container}`;
+}
+```
+
+
+#### Exercise 2: Visualizing the data
+
+It's time to put the Grammar of Graphics to good use. Create a scatterplot chart of the food supply and emissions data by following these steps:
+
+1. Instantiate the chart from the `df_food` dataframe
+2. Choose an appropriate graphical `mark_*()`
+3. Map `Total Food Supply (kg/capita/yr)` and `Emissions (Kg CO2 / kg product)` to visual encoding channels
+
+
+::: {.callout-tip icon="false" collapse="true"}
+
+# Solution
+```{python}
+
+alt.Chart(df_food).mark_point().encode(
+    x='Total Food Supply (kg/capita/yr)',
+    y='Emissions (Kg CO2 / kg product)',
+    tooltip='Food Item'
+)
+```
+
```

Comment:
Note that we will see if we can implement the AI advise here as well, or if we should create a "hint" button (i think maybe hints could help and give variety).

Reply:
Great idea. Yes ideally we would have LLM-powered hints that views the existing code.